File: app/src/main/java/com/smartpocket/musicwidget/activities/SongListActivity.java
package com.smartpocket.musicwidget.activities;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import com.futuremind.recyclerviewfastscroll.FastScroller;
import com.smartpocket.musicwidget.MusicWidget;
import com.smartpocket.musicwidget.R;
import com.smartpocket.musicwidget.backend.SongClickListener;
import com.smartpocket.musicwidget.backend.SongCursorRecyclerAdapter;
import com.smartpocket.musicwidget.backend.SongListLoader;
import com.smartpocket.musicwidget.model.Song;
import com.smartpocket.musicwidget.service.MusicService;

import org.jetbrains.annotations.NotNull;

public class SongListActivity extends AppCompatActivity implements SearchView.OnQueryTextListener {
    private final SongListLoader songLoader = SongListLoader.getInstance(this);
    private SongCursorRecyclerAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.song_list_activity);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(getResources().getString(R.string.title_activity_song_list));
        getSupportActionBar().setLogo(R.drawable.ic_launcher);

        handleIntent(getIntent());

        RecyclerView recyclerView = findViewById(R.id.listView);
        FastScroller fastScroller = findViewById(R.id.fastscroll);

        Cursor listCursor = songLoader.getCursor();
        adapter = new SongCursorRecyclerAdapter(listCursor, new SongClickListener() {
            @Override
            public void onSongSelected(@NotNull Song song) {
                Log.i("SongListactivity click", song.toString());
                Intent serviceIntent = new Intent(SongListActivity.this, MusicService.class);
                serviceIntent.setAction(MusicWidget.ACTION_JUMP_TO);
                serviceIntent.putExtra("song", song);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    SongListActivity.this.startForegroundService(serviceIntent);
                } else {
                    SongListActivity.this.startService(serviceIntent);
                }
                finish();
            }
        });

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                LinearLayout.VERTICAL);
        recyclerView.addItemDecoration(dividerItemDecoration);
        recyclerView.setAdapter(adapter);

        //has to be called AFTER RecyclerView.setAdapter()
        fastScroller.setRecyclerView(recyclerView);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.song_list, menu);
        MenuItem searchItem = menu.findItem(R.id.search);

        // Associate searchable configuration with the SearchView
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView = (SearchView) searchItem.getActionView();
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setOnQueryTextListener(this);

        return true;
    }


    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            String searchText = intent.getStringExtra(SearchManager.QUERY);
            Log.d("SongListActivity", "handleIntent Searched for: " + searchText);
            doSearch(searchText);
        }
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        Log.d("SongListActivity", "onQueryTextChange Searched for: " + newText);
        doSearch(newText);
        return true;
    }

    private void doSearch(String newText) {
        Cursor newCursor = songLoader.getFilteredCursor(newText);
        adapter.changeCursor(newCursor);
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }
}

File: app/src/main/java/com/smartpocket/musicwidget/activities/SongListActivity.kt
package com.smartpocket.musicwidget.activities

import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import com.smartpocket.musicwidget.MusicWidget
import com.smartpocket.musicwidget.R
import com.smartpocket.musicwidget.backend.AdViewHelper
import com.smartpocket.musicwidget.backend.SongClickListener
import com.smartpocket.musicwidget.backend.SongCursorRecyclerAdapter
import com.smartpocket.musicwidget.databinding.SongListActivityBinding
import com.smartpocket.musicwidget.model.Song
import com.smartpocket.musicwidget.service.MusicService
import org.koin.android.viewmodel.ext.android.viewModel
import java.util.concurrent.atomic.AtomicBoolean

private const val REQ_CODE = 1

class SongListActivity : AppCompatActivity(), SearchView.OnQueryTextListener {
    private lateinit var adapter: SongCursorRecyclerAdapter
    private lateinit var adViewHelper: AdViewHelper
    private val viewModel: SongListVM by viewModel()
    private var isSearching = AtomicBoolean(false)
    private lateinit var binding: SongListActivityBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = SongListActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)
        handleIntent(intent)

        adapter = SongCursorRecyclerAdapter(null, this, object : SongClickListener {
            override fun onSongSelected(song: Song) {
                Log.i("SongListActivity click", song.toString())
                val serviceIntent = Intent(this@SongListActivity, MusicService::class.java)
                serviceIntent.action = MusicWidget.ACTION_JUMP_TO
                serviceIntent.putExtra("song", song)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    startForegroundService(serviceIntent)
                } else {
                    startService(serviceIntent)
                }
                finish()
            }
        })

        binding.listView.also {
            val dividerItemDecoration = DividerItemDecoration(it.context, LinearLayout.VERTICAL)
            it.addItemDecoration(dividerItemDecoration)
            it.adapter = adapter
            it.addOnScrollListener(MyRecyclerViewOnScrollListener(binding.fabBtn, isSearching))

            //has to be called AFTER RecyclerView.setAdapter()
            with(binding.fastscroll) {
                setRecyclerView(it)
                setViewProvider(MyFastScrollScrollerViewProvider(binding.fabBtn))
            }
        }

        viewModel.cursorLD.observe(this, {
            adapter.changeCursor(it)
            binding.tvNoSongsFound.visibility = if (it == null) View.VISIBLE else View.GONE
        })
        viewModel.currentPosLD.observe(this, Observer(binding.listView::scrollToPosition))

        if (needsToRequestPermissions()) {
            val intent = Intent(this, ConfigurationActivity::class.java)
            startActivityForResult(intent, REQ_CODE)
        } else {
            viewModel.getCursor()
        }
        adViewHelper = AdViewHelper(binding.adViewContainer, this)
        adViewHelper.showBanner(true)
    }

    override fun onDestroy() {
        adViewHelper.destroy()
        super.onDestroy()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQ_CODE) {
            if (needsToRequestPermissions()) {
                finish()
            } else {
                viewModel.getCursor()
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.song_list, menu)
        val searchItem = menu.findItem(R.id.search)
        searchItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem?): Boolean {
                isSearching.set(true)
                binding.fabBtn.shrink()
                return true
            }

            override fun onMenuItemActionCollapse(item: MenuItem?): Boolean {
                isSearching.set(false)
                binding.fabBtn.extend()
                return true
            }
        })

        // Associate searchable configuration with the SearchView
        (searchItem.actionView as SearchView).also {
            val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
            it.setSearchableInfo(searchManager.getSearchableInfo(componentName))
            it.setOnQueryTextListener(this)
        }

        binding.fabBtn.setOnClickListener {
            if (searchItem.isActionViewExpanded) searchItem.collapseActionView()
            else searchItem.expandActionView()
        }
        return true
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            val searchText = intent.getStringExtra(SearchManager.QUERY)
            Log.d("SongListActivity", "handleIntent Searched for: $searchText")
            doSearch(searchText!!)
        }
    }

    override fun onQueryTextChange(newText: String): Boolean {
        Log.d("SongListActivity", "onQueryTextChange Searched for: $newText")
        doSearch(newText)
        return true
    }

    private fun doSearch(newText: String) = viewModel.getCursor(newText)

    override fun onQueryTextSubmit(query: String) = false
}